File: community/cypher/cypher-compiler-2.0/src/main/scala/org/neo4j/cypher/internal/compiler/v2_0/blackbuck/impl/operator/UnionOperator.scala
package org.neo4j.cypher.internal.compiler.v2_0.blackbuck.impl.operator

import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.slot.Pos
import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.rows.{RowPool, Cursor}
import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.sched.{Scheduler, Operator}

abstract class UnionOperator[P <: Pos[P], C <: Cursor[P, C], E](key: AnyRef, dest: Operator[P, C, E],
                                                                lhs: Operator[P, C, E],
                                                                rhs: Operator[P, C, E])
  extends FusingOperator[P, C, E](key, dest, lhs, rhs)
{
  override def newActivation(pool: RowPool[P, C], scheduler: Scheduler[P, C, E]): UnionActivation[P, C, E]
}

File: community/cypher/cypher-compiler-2.0/src/main/scala/org/neo4j/cypher/internal/compiler/v2_0/blackbuck/impl/operator/UnionOperator.java
//package org.neo4j.cypher.internal.compiler.v2_0.blackbuck.impl.operator;
//
//import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.rows.RowPool;
//import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.sched.*;
//import org.neo4j.cypher.internal.compiler.v2_0.blackbuck.api.rows.Cursor;
//
//public class UnionOperator<C extends Cursor<C>> extends FusingOperator<C>
//{
//    public UnionOperator( Object key, Operator<C> destination, Operator<C> lhs, Operator<C> rhs )
//    {
//        super( key, destination, lhs, rhs );
//    }
//
//    @Override
//    public void install( final RowPool<C> pool, Scheduler<C> scheduler ) {
//        final Object lhsKey = lhs().key();
//        final Object rhsKey = rhs().key();
//
//        Activation<C> activation = new Activation<C>() {
//            private final C output = pool.newCursor();
//
//            private boolean lhs = true;
//
//            @Override
//            public void activate( Router<C> router, Event<C> event )
//            {
//                if ( event.isFrom( lhsKey ) )
//                {
//                    if ( ! lhs )
//                    {
//                        throw new IllegalStateException( "Union received lhs chunk out of order" );
//                    }
//
//                    emit( router, event.cursor() );
//
//                    if ( event.isLast() )
//                    {
//                        lhs = false;
//
//                        output.rewind();
//                        while ( output.hasNext() )
//                        {
//                            emit( router, output.nextChunk() );
//                        }
//                    }
//                }
//                else if ( event.isFrom( rhsKey ) )
//                {
//                    if ( lhs )
//                    {
//                        event.cursor().appendTo( output );
//                    }
//                    else
//                    {
//                        emit( router, event.cursor() );
//                    }
//                }
//                else
//                {
//                    throw new IllegalStateException( "Unknown event" );
//                }
//            }
//        };
//
//        scheduler.register( lhsKey, activation );
//        scheduler.register( rhsKey, activation );
//    }
//}
